src/sfloader/fluid_instpatch.c

```diff
@@ -0,0 +1,667 @@
+{
+    fluid_instpatch_voice_user_data_t* dat = user_data;
+    
+    delete_fluid_sample(dat->sample);
+    ipatch_sample_store_cache_close(dat->sample_store);
+    FLUID_FREE(dat);
+}
+
+static IpatchSF2VoiceCache* convert_dls_to_sf2_instrument(fluid_instpatch_font_t *patchfont, IpatchDLS2Inst *item)
+{
+    IpatchConverter *conv;
+    IpatchSF2VoiceCache *cache;
+    int i, count;
+
+    /* create SF2 voice cache converter */
+    conv = ipatch_create_converter(G_OBJECT_TYPE(item), IPATCH_TYPE_SF2_VOICE_CACHE);
+
+    /* no SF2 voice cache converter for this item type? */
+    if(conv == NULL)
+    {
+        FLUID_LOG(FLUID_ERR, "Unable to find a voice cache converter for this type");
+        return NULL;
+    }
+
+    cache = ipatch_sf2_voice_cache_new(NULL, 0);
+
+    if(cache == NULL)
+    {
+        FLUID_LOG(FLUID_ERR, "Out of memory");
+        g_object_unref(conv);
+        return NULL;
+    }
+
+    ipatch_converter_add_input(conv, G_OBJECT(item));
+    ipatch_converter_add_output(conv, G_OBJECT(cache));
+
+    if(!ipatch_converter_convert(conv, NULL))
+    {
+        FLUID_LOG(FLUID_ERR, "Failed to convert DLS inst to SF2 voices");
+        g_object_unref(cache);
+        g_object_unref(conv);
+        return NULL;
+    }
+
+    g_object_unref(conv);
+    conv = NULL;
+
+    /* Use voice->user_data to close open cached stores */
+    cache->voice_user_data_destroy = fluid_instpatch_on_voice_user_data_destroy;
+
+    /* loop over voices and load sample data into RAM */
+    count = cache->voices->len;
+
+    for(i = 0; i < count; i++)
+    {
+        IpatchSF2Voice *voice = &g_array_index(cache->voices, IpatchSF2Voice, i);
+
+        if(!ipatch_sf2_voice_cache_sample_data(voice, NULL))
+        {
+            FLUID_LOG(FLUID_ERR, "ipatch_sf2_voice_cache_sample_data() failed");
+            g_object_unref(cache);
+            return NULL;
+        }
+        
+        if((voice->user_data = new_fluid_instpatch_voice_user_data(IPATCH_SAMPLE_STORE_CACHE(voice->sample_store))) == NULL)
+        {
+            g_object_unref(cache);
+            return NULL;
+        }
+    }
+
+    /* !! caller takes over cache reference */
+    return cache;
+}
+
+
+fluid_instpatch_font_t *new_fluid_instpatch(fluid_sfont_t *sfont, const fluid_file_callbacks_t *fcbs, const char *filename)
+{
+    fluid_instpatch_font_t *patchfont = NULL;
+    GError *err = NULL;
+    IpatchDLSReader *reader = NULL;
+    IpatchDLSFile *file = NULL;
+    IpatchFileHandle *handle = NULL;
+
+    fluid_return_val_if_fail(filename != NULL, NULL);
+
+    if((patchfont = FLUID_NEW(fluid_instpatch_font_t)) == NULL)
+    {
+        return NULL;
+    }
+
+    FLUID_MEMSET(patchfont, 0, sizeof(*patchfont));
+
+    FLUID_STRNCPY(&patchfont->name[0], filename, sizeof(patchfont->name));
+
+    /* open a file, we get a reference */
+    if((file = ipatch_dls_file_new()) == NULL)
+    {
+        FLUID_FREE(patchfont);
+
+        return NULL;
+    }
+
+    /* TODO
```

Comment:
Maybe add a short comment explaining why that is still a TODO and currently not really needed.

Reply:
Sry. That's a leftover, which I've dropped, see my first comment of this PR.